### ISSUE TITLE
Warning for assigning scalar to array

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#13025]
Migrated from [rt.perl.org#118469](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=118469) (status was 'rejected')

Searchable as RT118469$

**Comment by p5pRT** [comment]
### From @epa

#### Created by @epa



I was surprised this didn't generate a warning&#8203;:

&nbsp;    \#\!/usr/bin/perl
&nbsp;    use 5\.016;
&nbsp;    use warnings;
&nbsp;    use strict;
&nbsp;    my @&#8203;row = \[ 56 \];

In fact\, any scalar assignment happens without a warning&#8203;:

&nbsp;    my @&#8203;row = 3;

Since you can just as easily \(and more clearly\) say instead

&nbsp;    my @&#8203;row = \(3\);

then I think there should be a warning if a plain scalar
\(determined by syntax\, not at run time\) is assigned to an array\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=wishlist

Site configuration information for perl 5.16.3:

Configured by Red Hat, Inc. at Fri May  3 12:10:03 UTC 2013.

Summary of my perl5 (revision 5 version 16 subversion 3) configuration:
   
  Platform:
    osname=linux, osvers=2.6.32-358.2.1.el6.x86_64, archname=x86_64-linux-thread-multi
    uname='linux buildvm-24.phx2.fedoraproject.org 2.6.32-358.2.1.el6.x86_64 #1 smp wed feb 20 12:17:37 est 2013 x86_64 x86_64 x86_64 gnulinux '
    config_args='-des -Doptimize=-O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4  -m64 -mtune=generic -Dccdlflags=-Wl,--enable-new-dtags -Dlddlflags=-shared -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4  -m64 -mtune=generic -Wl,-z,relro  -DDEBUGGING=-g -Dversion=5.16.3 -Dmyhostname=localhost -Dperladmin=root@localhost -Dcc=gcc -Dcf_by=Red Hat, Inc. -Dprefix=/usr -Dvendorprefix=/usr -Dsiteprefix=/usr/local -Dsitelib=/usr/local/share/perl5 -Dsitearch=/usr/local/lib64/perl5 -Dprivlib=/usr/share/perl5 -Dvendorlib=/usr/share/perl5/vendor_perl -Darchlib=/usr/lib64/perl5 -Dvendorarch=/usr/lib64/perl5/vendor_perl -Darchname=x86_64-linux-thread-multi -Dlibpth=/usr/local/lib64 /lib64 /usr/lib64 -Duseshrplib -Dusethreads -Duseithreads -Dusedtrace=/usr/bin/dtrace -Duselargefiles -Dd_semctl_semun -Di_db -Ui_ndbm -Di_gdbm -Di_shadow -Di_syslog -Dman3ext=3pm -Duseperlio -Dinstallusrbinperl=n -Ubincompat5005 -Uversiononly -Dpager=/usr/bin/less -isr -Dd_gethostent_r_proto -Ud_endhostent_r_proto -Ud_sethostent_r_proto -Ud_endprotoent_r_proto -Ud_setprotoent_r_proto -Ud_endservent_r_proto -Ud_setservent_r_proto -Dscriptdir=/usr/bin -Dusesitecustomize'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=define, usemultiplicity=define
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=define, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='gcc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4 -m64 -mtune=generic',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include'
    ccversion='', gccversion='4.7.2 20121109 (Red Hat 4.7.2-8)', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='gcc', ldflags =' -fstack-protector'
    libpth=/usr/local/lib64 /lib64 /usr/lib64
    libs=-lresolv -lnsl -lgdbm -ldb -ldl -lm -lcrypt -lutil -lpthread -lc -lgdbm_compat
    perllibs=-lresolv -lnsl -ldl -lm -lcrypt -lutil -lpthread -lc
    libc=, so=so, useshrplib=true, libperl=libperl.so
    gnulibc_version='2.16'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,--enable-new-dtags -Wl,-rpath,/usr/lib64/perl5/CORE'
    cccdlflags='-fPIC', lddlflags='-shared -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4 -m64 -mtune=generic -Wl,-z,relro '

Locally applied patches:
    


@INC for perl 5.16.3:
    /home/eda/lib/perl5/
    /usr/local/lib64/perl5
    /usr/local/share/perl5
    /usr/lib64/perl5/vendor_perl
    /usr/share/perl5/vendor_perl
    /usr/lib64/perl5
    /usr/share/perl5
    .


Environment for perl 5.16.3:
    HOME=/home/eda
    LANG=en_GB.UTF-8
    LANGUAGE (unset)
    LC_COLLATE=C
    LC_CTYPE=en_GB.UTF-8
    LC_MESSAGES=en_GB.UTF-8
    LC_MONETARY=en_GB.UTF-8
    LC_NUMERIC=en_GB.UTF-8
    LC_TIME=en_GB.UTF-8
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/eda/bin:/home/eda/bin:/usr/local/bin:/usr/bin:/sbin:/usr/sbin:/sbin:/usr/sbin
    PERL5LIB=/home/eda/lib/perl5/
    PERL_BADLANG (unset)
    SHELL=/bin/bash


______________________________________________________________________
This email has been scanned by the Symantec Email Security.cloud service.
For more information please visit http://www.symanteccloud.com
______________________________________________________________________

```
</details>

**Comment by p5pRT** [comment]
### From @mauke

On 13\.06\.2013 14&#8203;:36\, Ed Avis \(via RT\) wrote&#8203;:
>
> I was surprised this didn't generate a warning&#8203;:
>
>      \#\!/usr/bin/perl
>      use 5\.016;
>      use warnings;
>      use strict;
>      my @&#8203;row = \[ 56 \];
>
> In fact\, any scalar assignment happens without a warning&#8203;:
>
>      my @&#8203;row = 3;
>
> Since you can just as easily \(and more clearly\) say instead
>
>      my @&#8203;row = \(3\);
>
> then I think there should be a warning if a plain scalar
> \(determined by syntax\, not at run time\) is assigned to an array\.

I don't like this suggestion because it clashes with my mental model of 
Perl\. You're not assigning a scalar to an array\, you're assigning a list 
\(containing one scalar\)\. And listiness is determined purely by context; 
there is no syntax for list literals or anything like that\.

There are some unfortunate special cases where parens force 
interpretation as a list \(namely with the assignment and repetition 
operators\, '\( \.\.\. \) = \.\.\.' and '\( \.\.\. \) x \.\.\.'\)\. But apart from that 
parens are entirely about precedence\. That's why I think perl shouldn't 
differentiate between '\(3\)' and '3'\.

Other opinions?


P\.S\.&#8203;: What exactly counts as a "plain scalar"? Should '@&#8203;x = 2 \+ 2' or 
'@&#8203;x = $y \. $z' warn? What about '@&#8203;x = 1 \.\. 10'?

P\.P\.S\.&#8203;: I assume this warning would apply to any instance of list 
assignment\, not just to arrays\.

\-\- 
Lukas Mai \<plokinom@&#8203;gmail\.com>

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @demerphq

On 13 June 2013 21&#8203;:01\, Lukas Mai \<plokinom@&#8203;gmail\.com> wrote&#8203;:
> On 13\.06\.2013 14&#8203;:36\, Ed Avis \(via RT\) wrote&#8203;:
>>
>>
>> I was surprised this didn't generate a warning&#8203;:
>>
>>      \#\!/usr/bin/perl
>>      use 5\.016;
>>      use warnings;
>>      use strict;
>>      my @&#8203;row = \[ 56 \];
>>
>> In fact\, any scalar assignment happens without a warning&#8203;:
>>
>>      my @&#8203;row = 3;
>>
>> Since you can just as easily \(and more clearly\) say instead
>>
>>      my @&#8203;row = \(3\);
>>
>> then I think there should be a warning if a plain scalar
>> \(determined by syntax\, not at run time\) is assigned to an array\.
>
>
> I don't like this suggestion because it clashes with my mental model of
> Perl\. You're not assigning a scalar to an array\, you're assigning a list
> \(containing one scalar\)\. And listiness is determined purely by context;
> there is no syntax for list literals or anything like that\.
>
> There are some unfortunate special cases where parens force interpretation
> as a list \(namely with the assignment and repetition operators\, '\( \.\.\. \) =
> \.\.\.' and '\( \.\.\. \) x \.\.\.'\)\.

There are a few more\. \(\) is the empty list\.

Such as&#8203;:

@&#8203;array= \(\);
$count = \(\) = $string =~ /\\d\+/g;

And there is list slice notation&#8203;:

\(1\,2\,3\)\[1\]
\(\)\[1\]

etc\. perldoc perldata has a bunch of discussion on it\.

> But apart from that parens are entirely about
> precedence\. That's why I think perl shouldn't differentiate between '\(3\)'
> and '3'\.
>
> Other opinions?

I dont agree with your reasons\, but I agree with the result\.

I think it is documented somewhere to work like this\, and I dont think
we could change even if we wanted to\.

Yves




\-\-
perl \-Mre=debug \-e "/just|another|perl|hacker/"

**Comment by p5pRT** [comment]
### From @mauke

On 13\.06\.2013 21&#8203;:28\, demerphq wrote&#8203;:
> On 13 June 2013 21&#8203;:01\, Lukas Mai \<plokinom@&#8203;gmail\.com> wrote&#8203;:
 >
>> There are some unfortunate special cases where parens force interpretation
>> as a list \(namely with the assignment and repetition operators\, '\( \.\.\. \) =
>> \.\.\.' and '\( \.\.\. \) x \.\.\.'\)\.
>
> There are a few more\. \(\) is the empty list\.
>
> Such as&#8203;:
>
> @&#8203;array= \(\);
> $count = \(\) = $string =~ /\\d\+/g;

Good point\. I'd argue it's slightly different because here the parens 
don't surround an expressions\, so it can't be confused with precedence\.

> And there is list slice notation&#8203;:
>
> \(1\,2\,3\)\[1\]
> \(\)\[1\]
>

IMO that doesn't count because here the parens are lexically part of the 
\( \)\[ \] operator\. In my head this is stored as a different kind of 
symbol\, i\.e\. an operator \(along with \->\( \)\, etc\.\)\, unlike plain parens\.


\-\- 
Lukas Mai \<plokinom@&#8203;gmail\.com>

**Comment by p5pRT** [comment]
### From @epa

It's not true that 'parens are entirely about precedence'\, at least not for list assignment\.
Usually you cannot omit the parens\.  Consider&#8203;:

&nbsp;    @&#8203;a = \(\);    \# OK
&nbsp;    @&#8203;a = ;      \# error

&nbsp;    @&#8203;a = \(1\, 2\);    \# OK
&nbsp;    @&#8203;a = 1\, 2;      \# compiles\, but probably not what you wanted

Now you might say that the second example is just the comma operator in action\.
Technically that is true\, but the comma operator is rarely used in everyday code\.
\(Perhaps only in the C\-style for loop and in the C\<warn\(\)\, next if $foo> idiom\.\)
It's the same in C\, the comma operator exists but most programmers are only
dimly aware of it and use it in a few stock phrases only\.

My point is that pretty universally in Perl code\, parens are used in a list assigment\.
That's what I learned starting with the llama book\, then the camel book\, and all
the way on through the manual pages and most CPAN code\.  Even after many
years it was almost surprising to see that naked assignment C\<@&#8203;a = 5> accepted\.

I don't suggest that it should stop being accepted \- it is too late to change\.
But the idea of a warning is to flag something which is technically part of the
language\, but uncommon and more likely than not to be a mistake\.  In
particular I think that

&nbsp;    @&#8203;a = \[ 1\, 2\, 3 \];

looks seductively like a list assignment\, especially for those who also program
in the many other languages that use \[\] syntax for lists\.  So I would make a
warning for that if for nothing else\.

\-\- 
Ed Avis \<eda@&#8203;waniasset\.com>

\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_
This email has been scanned by the Symantec Email Security\.cloud service\.
For more information please visit http&#8203;://www\.symanteccloud\.com
\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_

**Comment by p5pRT** [comment]
### From @mauke

On 13\.06\.2013 22&#8203;:26\, Ed Avis wrote&#8203;:
> It's not true that 'parens are entirely about precedence'\, at least not for list assignment\.
> Usually you cannot omit the parens\.  Consider&#8203;:
>
>      @&#8203;a = \(\);    \# OK
>      @&#8203;a = ;      \# error
>
>      @&#8203;a = \(1\, 2\);    \# OK
>      @&#8203;a = 1\, 2;      \# compiles\, but probably not what you wanted
>
> Now you might say that the second example is just the comma operator in action\.
> Technically that is true\, but the comma operator is rarely used in everyday code\.
> \(Perhaps only in the C\-style for loop and in the C\<warn\(\)\, next if $foo> idiom\.\)
> It's the same in C\, the comma operator exists but most programmers are only
> dimly aware of it and use it in a few stock phrases only\.

No\, I'm saying both '@&#8203;a = \(1\, 2\)' and '@&#8203;a = 1\, 2' are the comma operator 
in action\. In scalar context it evaluates its left operand\, ignores the 
result\, and returns the result of the right operand\. In list context it 
evaluates both operands in list context and concatenates them\.
TL;DR&#8203;: Comma in list context is list concatenation\.

In '@&#8203;a = \(1\, 2\)' the parens are just for precedence\. The assignment to 
@&#8203;a puts '1\, 2' in list context\. Both '1' and '2' are evaluated in list 
context\, yielding two lists of one element each\. Then '\,' concatenates 
them\, returning the two element list \<1 2>\.

> My point is that pretty universally in Perl code\, parens are used in a list assigment\.
> That's what I learned starting with the llama book\, then the camel book\, and all
> the way on through the manual pages and most CPAN code\.  Even after many
> years it was almost surprising to see that naked assignment C\<@&#8203;a = 5> accepted\.

> I don't suggest that it should stop being accepted \- it is too late to change\.
> But the idea of a warning is to flag something which is technically part of the
> language\, but uncommon and more likely than not to be a mistake\.  In
> particular I think that
>
>      @&#8203;a = \[ 1\, 2\, 3 \];
>
> looks seductively like a list assignment\, especially for those who also program
> in the many other languages that use \[\] syntax for lists\.  So I would make a
> warning for that if for nothing else\.

It is list assignment \- it assigns a list containing one element \(a 
reference\) to @&#8203;a\. You say that's probably not what the programmer 
intended\. But to me it seems like a completely natural thing to do\.

I grepped the perl directory for '@&#8203;\\w\+\\s\*=\\s\*\\\[' and found one instance&#8203;:
t/comp/parser\.t&#8203;:444&#8203;:     @&#8203;x = \['string'\];

And two in my directory of scratch programs&#8203;:
&nbsp;     @&#8203;tmp = \[\];
\.\.\.
&nbsp;     my @&#8203;stack = \[$from\, $cur\];


\(and one occurrence in File&#8203;::Copy&#8203;::Recursive \(line 85\) that's almost 
certainly a bug&#8203;:

&nbsp;         my @&#8203;call = \[$do\->\($path\, @&#8203;args\)\] or return;
&nbsp;         push @&#8203;rt\, \\@&#8203;call;

I'm not sure what the author was trying to do there\.\)

Anyway\, if this does turn into a warning\, I'm going to have to change 
some code\.

\-\- 
Lukas Mai \<plokinom@&#8203;gmail\.com>

**Comment by p5pRT** [comment]
### From @rjbs

I sympathize with the fact that @&#8203;x=\[1\,2\,3\] can be done in error\, but I think the cure would be 
worse than the disease\.  It's perfectly good Perl code\, not something best left for experts\, and 
adding a warning would be more nuisance than help\.  It's not going to happen\.

 
\-\- 
rjbs

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

I sympathize with the fact that @&#8203;x=\[1\,2\,3\] can be done in error\, but I think the cure would be 
worse than the disease\.  It's perfectly good Perl code\, not something best left for experts\, and 
adding a warning would be more nuisance than help\.  It's not going to happen\.

 
\-\- 
rjbs

**Comment by p5pRT** [comment]
@rjbs - Status changed from 'open' to 'rejected'

**Comment by p5pRT** [comment]
### From @ikegami

On Thu\, Jun 13\, 2013 at 3&#8203;:01 PM\, Lukas Mai \<plokinom@&#8203;gmail\.com> wrote&#8203;:

> On 13\.06\.2013 14&#8203;:36\, Ed Avis \(via RT\) wrote&#8203;:
>
>>
>> I was surprised this didn't generate a warning&#8203;:
>>
>>      \#\!/usr/bin/perl
>>      use 5\.016;
>>      use warnings;
>>      use strict;
>>      my @&#8203;row = \[ 56 \];
>>
>> In fact\, any scalar assignment happens without a warning&#8203;:
>>
>>      my @&#8203;row = 3;
>>
>> Since you can just as easily \(and more clearly\) say instead
>>
>>      my @&#8203;row = \(3\);
>>
>> then I think there should be a warning if a plain scalar
>> \(determined by syntax\, not at run time\) is assigned to an array\.
>>
>
> I don't like this suggestion because it clashes with my mental model of
> Perl\. You're not assigning a scalar to an array\, you're assigning a list
> \(containing one scalar\)\. And listiness is determined purely by context;
> there is no syntax for list literals or anything like that\.
>
> There are some unfortunate special cases where parens force interpretation
> as a list \(namely with the assignment and repetition operators\, '\( \.\.\. \) =
> \.\.\.' and '\( \.\.\. \) x \.\.\.'\)\. But apart from that parens are entirely about
> precedence\. That's why I think perl shouldn't differentiate between '\(3\)'
> and '3'\.
>

Parens aren't entirely about precedence\, but they never \(directly\) create a
list\. So why should they be required here? It makes no sense to whatsoever\.

It seems to me that the OP is trying to promote the idea that there is some
distinction between list of scalars and scalars that simply doesn't exist\.
Projecting his misconception unto other developers would do them a big
disservice\.

\- Eric

**Comment by p5pRT** [comment]
### From @ikegami

On Thu\, Jun 13\, 2013 at 4&#8203;:26 PM\, Ed Avis \<eda@&#8203;waniasset\.com> wrote&#8203;:

> I don't suggest that it should stop being accepted \- it is too late to
> change\.
>
But the idea of a warning is to flag something which is technically part of
> the
> language\, but uncommon and more likely than not to be a mistake\.  In
> particular I think that
>
>     @&#8203;a = \[ 1\, 2\, 3 \];
>

I wouldn't have a problem with warning for this particular case\. It's a
frequent newcomer mistake\.

**Comment by p5pRT** [comment]
### From @ikegami

On Thu\, Jun 13\, 2013 at 5&#8203;:57 PM\, Eric Brine \<ikegami@&#8203;adaelis\.com> wrote&#8203;:

> On Thu\, Jun 13\, 2013 at 4&#8203;:26 PM\, Ed Avis \<eda@&#8203;waniasset\.com> wrote&#8203;:
>
>> I don't suggest that it should stop being accepted \- it is too late to
>> change\.
>>
> But the idea of a warning is to flag something which is technically part
>> of the
>> language\, but uncommon and more likely than not to be a mistake\.  In
>> particular I think that
>>
>>     @&#8203;a = \[ 1\, 2\, 3 \];
>>
>
> I wouldn't have a problem with warning for this particular case\. It's a
> frequent newcomer mistake\.
>

On the other hand\, @&#8203;x = \[1\,2\,3\]; doesn't tend to lead to subtle errors\.
When this error is present in the code of newcomers\, it's my experience
that the program fails quite noticeably\.

**Comment by p5pRT** [comment]
### From j.imrie1@virginmedia.com

On 13/06/2013 22&#8203;:57\, Eric Brine wrote&#8203;:
> On Thu\, Jun 13\, 2013 at 4&#8203;:26 PM\, Ed Avis \<eda@&#8203;waniasset\.com 
> \<mailto&#8203;:eda@&#8203;waniasset\.com>> wrote&#8203;:
>
>     I don't suggest that it should stop being accepted \- it is too
>     late to change\.
>
>     But the idea of a warning is to flag something which is
>     technically part of the
>     language\, but uncommon and more likely than not to be a mistake\.  In
>     particular I think that
>
>         @&#8203;a = \[ 1\, 2\, 3 \];
>
>
> I wouldn't have a problem with warning for this particular case\. It's 
> a frequent newcomer mistake\.
>
Could we have a new warning category for  newcomer mistakes that is 
turned off by default but can be activated with

use warnings 'newcomer';

**Comment by p5pRT** [comment]
### From @Leont

On Thu\, Jun 13\, 2013 at 2&#8203;:36 PM\, Ed Avis \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> I was surprised this didn't generate a warning&#8203;:
>
>     \#\!/usr/bin/perl
>     use 5\.016;
>     use warnings;
>     use strict;
>     my @&#8203;row = \[ 56 \];
>
> In fact\, any scalar assignment happens without a warning&#8203;:
>
>     my @&#8203;row = 3;
>
> Since you can just as easily \(and more clearly\) say instead
>
>     my @&#8203;row = \(3\);
>
> then I think there should be a warning if a plain scalar
> \(determined by syntax\, not at run time\) is assigned to an array\.

I think this is the sort of thing that better suited for perlcritic
than a core warning\.

Leon

**Comment by p5pRT** [comment]
### From @epa

Eric Brine \<ikegami \<at> adaelis\.com> writes&#8203;:

>It seems to me that the OP is trying to promote the idea that there is some
>distinction between list of scalars and scalars that simply doesn't exist\.

You may be right\.  In that case\, it's a pity that

&nbsp;    @&#8203;a = 1\, 2\, 3;

does not assign a list of three elements\.  But it is probably too late to
change the operator precedence now\.

\-\- 
Ed Avis \<eda@&#8203;waniasset\.com>

**Comment by p5pRT** [comment]
### From Eirik-Berg.Hanssen@allverden.no

On Fri\, Jun 14\, 2013 at 10&#8203;:26 AM\, Ed Avis \<eda@&#8203;waniasset\.com> wrote&#8203;:

> Eric Brine \<ikegami \<at> adaelis\.com> writes&#8203;:
>
> >It seems to me that the OP is trying to promote the idea that there is
> some
> >distinction between list of scalars and scalars that simply doesn't exist\.
>
> You may be right\.  In that case\, it's a pity that
>
>     @&#8203;a = 1\, 2\, 3;
>
> does not assign a list of three elements\.  But it is probably too late to
> change the operator precedence now\.
>

&nbsp;  Yeah\, the most you could do now is add a low\-precedence assignment
operator&#8203;:

&nbsp;  @&#8203;a ass 1\, 2\, 3;

&nbsp;  \{g\}


Eirik

**Comment by p5pRT** [comment]
### From @mauke

On 14\.06\.2013 00&#8203;:58\, John Imrie wrote&#8203;:
> Could we have a new warning category for  newcomer mistakes that is
> turned off by default but can be activated with
>
> use warnings 'newcomer';

I think this approach is fundamentally flawed because it requires opt\-in 
from newcomers\, i\.e\. exactly the people who aren't quite sure what \(not\) 
to do\.

\(And of course you can't just start every tutorial with "use warnings 
'newcomer';" because then people not on modern perls \(i\.e\. people using 
their distro perl and just wanting to get started\) won't be able to run 
the examples\.\)

I think the 'use strictures' approach \(default to extra checks\, make 
experts opt out\) is technically ugly but fundamentally right because 
effective\.

\-\- 
Lukas Mai \<plokinom@&#8203;gmail\.com>
